File: CommunitySite.CodeSamples/Controllers/SampleController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Web.Hosting;
using System.Web.Http;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace CommunitySite.CodeSamples.Controllers
{
    public class SampleController : ApiController
    {
        [Route("api/samples")]
        public HttpResponseMessage GetByTopic(string topic)
        {
            var topicPath = string.Empty;

            try
            {
                if (string.IsNullOrWhiteSpace(topic))
                {
                    throw new ArgumentNullException("Please specify a valid topic parameter in the form of /sample/<topic>.");
                }
                topicPath = HostingEnvironment.MapPath("~/" + topic);
                if (!Directory.Exists(topicPath))
                {
                    throw new FileNotFoundException("Could not find a topic named: " + topic);
                }
            }
            catch (ArgumentNullException ex)
            {
                return CreateErrorResponse(ex);
            }
            catch (FileNotFoundException ex)
            {
                return CreateErrorResponse(ex);
            }

            var directories = new DirectoryInfo(topicPath);
            var jArray = new JArray();

            foreach (var langDir in directories.GetDirectories())
            {
                var examples = new JArray();
                var files = langDir.GetFiles();
                var exampleFiles = files.Where(f => f.Extension != ".json");
                var rootObject = new JObject();
                rootObject["mode"] = langDir.Name;
                rootObject["content"] = string.Empty;
                rootObject["id"] = langDir.Name;
                var filePathRoot = "https://v1codesamples.azurewebsites.net/api/sample?path=" + topic + "/" + langDir.Name + "/";

                foreach (var file in exampleFiles)
                {
                    var fileNameWithoutExtension = Path.GetFileNameWithoutExtension(file.Name);
                    var properties = files.First(f => f.Name == fileNameWithoutExtension + ".json");
                    using (var reader = new StreamReader(properties.OpenRead()))
                    {
                        var json = reader.ReadToEnd();
                        var exampleRoot = rootObject.DeepClone() as JObject;
                        var mergeWith = JObject.Parse(json);
                        var example = new JObject(exampleRoot.Concat(mergeWith.AsJEnumerable()));
                        example["url"] = filePathRoot + file.Name;
                        // TODO: multiple examples per language support
                        //examples.Add(example);
                        jArray.Add(example);
                    }
                }
                // TODO: support multiple examples
                //jArray.Add(examples);
            }

            var content = jArray.ToString();

            var resp = new HttpResponseMessage(HttpStatusCode.OK);
            resp.Content = new StringContent(content, Encoding.UTF8, "text/plain");
            return resp;

        }

        public HttpResponseMessage Get(string path)
        {
            var samplePath = string.Empty;

            try
            {
                if (string.IsNullOrWhiteSpace(path))
                {
                    throw new ArgumentNullException("Please specify a valid path parameter in the form of /sample/<path>.");
                }
                samplePath = HostingEnvironment.MapPath("~/" + path);
                if (!File.Exists(samplePath))
                {
                    throw new FileNotFoundException("Could not find file at the specified path: " + path);
                }
            }
            catch (ArgumentNullException ex) 
            {
                return CreateErrorResponse(ex);
            }
            catch (FileNotFoundException ex)
            {
                return CreateErrorResponse(ex);
            }

            var content = new StreamReader(samplePath).ReadToEnd();
            var resp = new HttpResponseMessage(HttpStatusCode.OK);
            resp.Content = new StringContent(content, Encoding.UTF8, "text/plain");
            return resp;
        }

        private static HttpResponseMessage CreateErrorResponse(Exception ex)
        {
            var resp = new HttpResponseMessage(HttpStatusCode.InternalServerError);
            resp.Content = new StringContent(ex.Message);
            return resp;
        }
    }
}

File: CommunityService/SampleController.fs
namespace CommunityService

open System
open System.Web.Http
open System.IO
open System.Net.Http
open System.Net
open System.Text
open Newtonsoft.Json
open Newtonsoft.Json.Linq
open System.Collections.Generic

type HttpRouteDefaults = { Controller : string; Id : obj }

type Global() = 
    inherit System.Web.HttpApplication()
    member this.Application_Start (sender: obj) (e: EventArgs) =
        GlobalConfiguration.Configuration.MapHttpAttributeRoutes() |> ignore
        GlobalConfiguration.Configuration.Routes.MapHttpRoute(
            "DefaultAPI", 
            "api/{controller}/{id}",
            { Controller = "Home"; Id = RouteParameter.Optional }) |> ignore
        GlobalConfiguration.Configuration.EnsureInitialized() |> ignore

module CommunityAPI =
    let CreateErrorResponse errorMessage =
        let response = new HttpResponseMessage(HttpStatusCode.InternalServerError)
        response.Content <- new StringContent(errorMessage)
        response

    let str value =
        JValue.CreateString value

    type SampleController() =
        inherit ApiController()
        member this.Get path =
            let samplePath = System.Web.Hosting.HostingEnvironment.MapPath("~/" + path)
            if String.IsNullOrWhiteSpace path then CreateErrorResponse "Please specify a valid path parameter in the form of /sample/<path>."
            elif File.Exists samplePath = false then CreateErrorResponse ("Could not find file at the specified path: " + path)
            else 
                use stream = new StreamReader(samplePath)
                let content = stream.ReadToEnd()
                let response = new HttpResponseMessage(HttpStatusCode.OK)
                response.Content <- new StringContent(content, Encoding.UTF8, "text/plain")
                response

        [<Route("api/samples")>]
        member this.GetByTopic topic =
            let topicPath = System.Web.Hosting.HostingEnvironment.MapPath("~/" + topic)
            if String.IsNullOrWhiteSpace topicPath then CreateErrorResponse "Please specify a valid path parameter in the form of /sample?topic=<topic>."
            elif Directory.Exists topicPath = false then CreateErrorResponse ("Could not find a topic named: " + topic)
            else 
                let directory = DirectoryInfo(topicPath)
                let topicExamples = JArray()
                for langDir in directory.GetDirectories() do
                    // let langExamples = new JArray() // TODO: support multiple examples per language
                    let files = langDir.GetFiles()
                    let exampleFiles = query {
                        for file in files do 
                        where (file.Extension <> ".json")
                        select file
                    }

                    let rootObject = JObject()
                    rootObject.["mode"] <- str langDir.Name
                    rootObject.["id"] <- str langDir.Name
                    rootObject.["content"] <- str String.Empty

                    let filePathRoot = "https://v1codesamples.azurewebsites.net/api/sample?path=" + topic + "/" + langDir.Name + "/"
                    for file in exampleFiles do
                        let fileNameWithoutExtension = Path.GetFileNameWithoutExtension(file.Name)
                        let properties = query { 
                            for file in files do
                            where (file.Name = fileNameWithoutExtension + ".json")
                            head
                        }
                        use reader = new StreamReader(properties.OpenRead())
                        let json = reader.ReadToEnd()
                        let exampleRoot = rootObject.DeepClone() :?> JObject;
                        let mergeWith = JObject.Parse(json)
                        let example = JObject(exampleRoot)
                        for token in mergeWith.Properties() do example.Add(token.Name, token.Value)
                        example.["url"] <- str (filePathRoot + file.Name)                        
                        //langExamples.Add(example); // TODO: multiple examples per language support
                        topicExamples.Add(example)

                let content = topicExamples.ToString()

                let response = new HttpResponseMessage(HttpStatusCode.OK);
                response.Content <- new StringContent(content, Encoding.UTF8, "text/plain");
                response